responsemanager/responseassembler/responseassembler.go

```diff
@@ -0,0 +1,112 @@
+/*
+Package responseassembler assembles responses that are queued for sending in outgoing messages
+
+The response assembler's Transaction method allows a caller to specify response actions that will go into a single
+libp2p2 message. The response assembler will also deduplicate blocks that have already been sent over the network in
+a previous message
+*/
+package responseassembler
+
+import (
+	"context"
+
+	"github.com/ipld/go-ipld-prime"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	"github.com/ipfs/go-graphsync"
+	gsmsg "github.com/ipfs/go-graphsync/message"
+	"github.com/ipfs/go-graphsync/notifications"
+	"github.com/ipfs/go-graphsync/peermanager"
+)
+
+// Transaction is a series of operations that should be send together in a single response
+type Transaction func(PeerResponseTransactionBuilder) error
+
+// PeerResponseTransactionBuilder is interface for assembling responses inside a transaction, so that they are included
+// in the same message on the protocol
+type PeerResponseTransactionBuilder interface {
+	SendResponse(
```

Comment:
These really need comments to explain how this interface is used.
```suggestion
	// SendResponse adds a response to this transaction.
	SendResponse(
```